File: app/src/main/java/com/michaelflisar/tests/MainActivity.java
package com.michaelflisar.tests;

import android.content.ComponentName;
import android.content.Intent;
import android.content.pm.ActivityInfo;
import android.content.pm.PackageManager;
import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.view.View;
import android.widget.TextView;

import com.michaelflisar.tests.databinding.MainActivityBinding;

public class MainActivity extends AppCompatActivity implements View.OnClickListener {
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        MainActivityBinding binding = DataBindingUtil.setContentView(this, R.layout.main_activity);

        int activities = createSubActivityButtons(binding);
        binding.tvInfo.setText(String.format("Existing MCVE Tests: %d", activities));
    }

    private int createSubActivityButtons(MainActivityBinding binding) {
        int activities = 0;
        try {
            ActivityInfo[] list = getPackageManager().getPackageInfo(getPackageName(), PackageManager.GET_ACTIVITIES).activities;
            activities = list.length - 1;
            for (ActivityInfo info : list) {
                if (!info.name.equals(MainActivity.class.getName())) {
                    TextView tv = (TextView) getLayoutInflater().inflate(R.layout.main_activity_list_item, binding.llActivities, false);
                    tv.setTag(info);
                    tv.setText(info.loadLabel(getPackageManager()));
                    tv.setOnClickListener(this);
                    binding.llActivities.addView(tv);
                }
            }
        } catch (PackageManager.NameNotFoundException e) {
            e.printStackTrace();
        }
        return activities;
    }

    @Override
    public void onClick(View v) {
        // no security checks, we know what we get
        ActivityInfo info = (ActivityInfo) v.getTag();
        ComponentName name = new ComponentName(info.packageName, info.name);
        Intent intent = new Intent(Intent.ACTION_MAIN);
        intent.setComponent(name);
        startActivity(intent);
    }
}

File: app/src/main/java/com/michaelflisar/tests/MainActivity.kt
package com.michaelflisar.tests

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.michaelflisar.tests.databinding.ActivityMainBinding

class MainActivity : AppCompatActivity(), View.OnClickListener {

    private lateinit var binding: ActivityMainBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(LayoutInflater.from(this))
        setSupportActionBar(binding.toolbar)
        setContentView(binding.root)

        addHeader("Functions (${Definitions.TEST_CASES.count { it is Definitions.TestCase.Function }})")
        createFunctionTestButtons()
        addHeader("Fragments (${Definitions.TEST_CASES.count { it is Definitions.TestCase.Fragment }})")
        createFragmentTestButtons()

        binding.tvInfo.text = "Existing MCVE Tests: ${Definitions.TEST_CASES.size}"
    }

    private fun createFunctionTestButtons() {
        Definitions.TEST_CASES.filterIsInstance<Definitions.TestCase.Function>().forEach {
            addItem(it.name, Definitions.TEST_CASES.indexOf(it))
        }
    }

    private fun createFragmentTestButtons() {
        Definitions.TEST_CASES.filterIsInstance<Definitions.TestCase.Fragment>().forEach {
            addItem(it.name, Definitions.TEST_CASES.indexOf(it))
        }
    }

    private fun addHeader(label: String) {
        val tv = layoutInflater.inflate(R.layout.item_test_case_header, binding.llTests, false) as TextView
        tv.text = label
        binding.llTests.addView(tv)
    }

    private fun addItem(label: String, tag: Any) {
        val tv = layoutInflater.inflate(R.layout.item_test_case, binding.llTests, false) as TextView
        tv.tag = tag
        tv.text = label
        tv.setOnClickListener(this)
        binding.llTests.addView(tv)
    }


    override fun onClick(v: View) {
        val index = v.tag as Int
        val testCase = Definitions.TEST_CASES[index]
        testCase.runTest(this)
        if (testCase is Definitions.TestCase.Function)
            Toast.makeText(this, "Function executed: ${testCase.name}", Toast.LENGTH_SHORT).show()
    }
}